### ISSUE TITLE
Add `transform` keyword to dataset constructors to store a transformation

### ISSUE THREAD

**Issue by CarloLucibello** [JuliaML/MLDatasets.jl#138]
Both [HuggingFace](https://huggingface.co/docs/datasets/v2.2.1/en/package_reference/main_classes#datasets.Dataset.set_transform) and torchvision's datasets can store a transformation that is applied on the fly when indexing a dataset. 
I think this is generally very convenient and we should add it to every dataset.

**Comment by lorenzoh** [comment]
I feel it may be cleaner to tell users to use `mapobs` and give some examples. AFAICT there would be no functional difference, right?

**Comment by CarloLucibello** [comment]
Yes maybe there is no advantage over a `mapobs`. I'll keep the issue open until we add some examples in the docs.

**Comment by CarloLucibello** [comment]
The recently introduced TorchData could be a source of inspiration
https://pytorch.org/data/beta/index.html
